### ISSUE TITLE
Remove KQL support if this code will not be fixed.

### ISSUE THREAD

**Issue by alexey-milovidov** [ClickHouse/ClickHouse#61742]
We have this ridiculous code in KQL:
```
bool isValidKQLPos(IParser::Pos & pos)
{
    return (pos.isValid() ||
            pos->type == TokenType::ErrorSingleExclamationMark || // allow kql negative operators
            pos->type == TokenType::ErrorWrongNumber || // allow kql timespan data type with decimal like 2.6h
            std::string_view(pos->begin, pos->end) == "~");  // allow kql Case-Sensitive operators
}
```
It first tokenizes a query as something containing errors. Then interprets errors as non-errors.
This should not be allowed. Let's quickly fix it or remove KQL entirely.

**Comment by alexey-milovidov** [comment]
Yes, this strange code exists only for Kusto.

**Comment by kashwy** [comment]
> Yes, this strange code exists only for Kusto.

I mean, this could be allowed for kql ? becasue  Then interprets errors as non-errors only for kql , which kql need this as `non-errors `

**Comment by kashwy** [comment]
> > Yes, this strange code exists only for Kusto.
> 
> I mean, this could be allowed for kql ? becasue Then interprets errors as non-errors only for kql , which kql need this as `non-errors `
this function is used for splitting kql statements to  kql pipes ,  if there's an error inside  kql pipe, it is handled

**Comment by alexey-milovidov** [comment]
```
Alexey Milovidov, [22-03-2024 16:50]
I found it while working on https://github.com/ClickHouse/ClickHouse/pull/61500

Everything was alright except KQL.
I know how to workaround it, but it looks strange.

Maybe introduce a flag, e.g. greedy_parse_errors for Tokens, and use it in KQL?

Yong Wang, [22-03-2024 16:52]
Thank you very much, I will check it
```

**Comment by kashwy** [comment]
thanks, working on it

**Comment by yusufozturk** [comment]
@kashwy just curiosity, how are you planning to support json field search, json extraction etc with kql on CH? Or is that something already implemented?

Are you going to use it in IBM Cloud Logs or QRadar Cloud? 

Sorry for my curiosity. I just find it very strange to see that kql is being implemented by IBM 😅 But yeah, it also makes sense actually.

**Comment by alexey-milovidov** [comment]
I've checked the KQL implementation and I'm slightly disappointed by its quality.
We can keep it under the experimental flag and discourage people from using it... but it isn't a good plan.

**Comment by kashwy** [comment]
The initial approach involved re-using the ClickHouse lexer, tokenizer, and parser to process KQL, primarily due to perceived similarities between KQL and SQL. This decision was based on leveraging existing, robust parsing mechanisms to accelerate development. However,  the nuanced differences between KQL and SQL have led to some issues that impact the overall quality of the implementation.
Our team is actively working on developing a dedicated KQL parser to better cater to its unique syntax and features. The first step in this process involves creating a specialized tokenizer for KQL. This will lay the groundwork for accurately recognizing KQL syntax elements and ensuring they are correctly interpreted. 
Following the tokenizer, our next step will be to develop a full expression parser specifically for KQL. This parser will be designed from the ground up to understand and efficiently process KQL expressions, thereby significantly improving the quality and reliability of the KQL implementation.
Our commitment to improving the KQL implementation extends beyond resolving the current issues. We are actively working on adding more features to ensure that users have a powerful and comprehensive tool at their disposal. This includes enhancing the existing capabilities and introducing new functionalities that are essential for a wide range of use cases.
We believe that by taking these steps to separate the KQL parser from the SQL parser, we can achieve a much higher quality solution that meets the needs of users. we look forward to delivering an improved KQL implementation in the near future. 
Your understanding and support during this process are greatly appreciated.

**Comment by alexey-milovidov** [comment]
Thank you, this is really appreciated!

**Comment by kashwy** [comment]
quick update, a PR for KQL with separated parser and tokenizer will be ready soon

**Comment by kashwy** [comment]
PR created

https://github.com/ClickHouse/ClickHouse/pull/62668

this is to isolate KQL parser from SQL parser, next  we will create more KQL specified parser to improve the parsing performance

**Comment by yusufozturk** [comment]
IBM is killing Qradar SaaS and apparently on-prem version will continue to get updates for a while which is not CH based. Will you continue to invest in Kusto support? @kashwy
/cc: @alexey-milovidov

**Comment by kashwy** [comment]
> IBM is killing Qradar SaaS and apparently on-prem version will continue to get updates for a while which is not CH based. Will you continue to invest in Kusto support? @kashwy /cc: @alexey-milovidov

Yes, there have been some recent changes at IBM, but we still have developers working on it and plan to continue providing support

**Comment by rschu1ze** [comment]
Is IBM still working on KQL?

**Comment by alexey-milovidov** [comment]
After I asked to rewrite it, they promptly did: https://github.com/ClickHouse/ClickHouse/pull/62668

**Comment by alexey-milovidov** [comment]
I checked what they did, and I'm horrified by the quality of this code.

**Comment by alexey-milovidov** [comment]
Fixed.
